source/extensions/common/wasm/context.cc

```diff
@@ -1181,17 +1187,23 @@ Http::FilterHeadersStatus Context::onRequestHeaders() {
   return Http::FilterHeadersStatus::StopIteration;
 }
 
-Http::FilterDataStatus Context::onRequestBody(int body_buffer_length, bool end_of_stream) {
+Http::FilterDataStatus Context::onRequestBody(int, bool end_of_stream) {
   if (!wasm_->on_request_body_) {
     return Http::FilterDataStatus::Continue;
   }
+  if (!end_of_stream_ && buffering_body_) {
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
```

Comment:
This forces consumers to select between streaming or waiting for the complete response.

However, this won't work, because large body (i.e. larger than the configured buffer limit) will get pushback, and the complete response will never arrive.

Also, there is a lot of use cases, when the consumer will want to wait until there are X bytes available (e.g. to process `<head>` section of the HTML body, or to get `PROXY-PROTOCOL` header in TCP, etc.), but not necessarily until the complete body is available.

As such, I think that we need to wake the consumer and call `proxy_on_request_body()` for each chunk, passing the size of the all available data, so that it can make a decision at each point.

We could possibly add `StopAndBufferUntilEndOfStream` if needed, but that's more of an optimization that would work only in some cases, than a generic solution.

Reply:
For bodies larger than the buffer limit, it's true that the callback will never arrive because the HTTP machinery will reject the request. But this won't help if someone wants to only read the first X bytes. So let me go back to calling the callback every time and documenting the effect.